### PR TITLE
Cirrus: enable Fedora 36 aarch64 tasks on EC2

### PR DESCRIPTION
**Pull request by lsm5**
Signed-off-by: Lokesh Mandvekar <lsm5@fedoraproject.org>

#### Does this PR introduce a user-facing change?

```release-note
None
```

Piggybacks on https://github.com/containers/podman/pull/14569

### CODE REVIEW

File: .cirrus.yml
Comment (cevich) [reply .cirrus.yml:219]: Yep, `matrix: *platform_axis` this won't work (and probably isn't what you want since there's no point in a single-axis matrix.

```
...
    ec2_instance: &standard_build_ec2_aarch64
      image: ${FEDORA_AARCH64_AMI_ID}
      type: t4g.micro
      region: us-east-1
      architecture: arm64
    env:
        <<: *stdenvars
        DISTRO_NV: ${FEDORA_NAME_AARCH64}
        TEST_FLAVOR: build
    clone_script: *full_clone
...
```

File: .cirrus.yml
Comment (cevich) [inline .cirrus.yml:202]: For consistency, this should probably be `${VM_IMAGE_NAME}`

File: .cirrus.yml
Comment (cevich) [inline .cirrus.yml:209]: For consistency, this should probably say `DISTRO_NV: ${FEDORA_NAME}` and `VM_IMAGE_NAME: ${FEDORA_NAME_AARCH64}`

File: .cirrus.yml
Comment (cevich) [inline .cirrus.yml:332]: Might need to add `VM_IMAGE_NAME: ${FEDORA_NAME_AARCH64}` here.

**Review by cevich — COMMENTED** [review]
Looking good so far

File: .cirrus.yml
Comment (cevich) [inline .cirrus.yml:279]:
```suggestion
-    clone_script: *get_gosrc
    # N/B: This script depends on ${DISTRO_NV} being defined for the task.
    clone_script: &get_gosrc_aarch64 |
        cd /tmp
        echo "$ARTCURL/build_aarch64/repo/repo.tbz"
        time $ARTCURL/build_aarch64/repo/repo.tbz
        time tar xjf /tmp/repo.tbz -C $GOSRC
```

That way you can use the alias `*get_gosrc_aarch64` in subsequent tasks

File: .cirrus.yml
Comment (cevich) [inline .cirrus.yml:335]:
```suggestion
    clone_script: *get_gosrc_aarch64
```

**Comment by lsm5** [comment]
@vrothberg any idea whatsup with `Validate fedora-36 {arm64} Build` ?

**Comment by lsm5** [comment]
ha, it was just this stupid `arm64 v/s aarch64` thing for ec2 arch specification. gotta be super careful doing a search and replace.

File: .cirrus.yml
Comment (cevich) [inline .cirrus.yml:255]: Why this?  Seems unrelated to this PR.

File: .cirrus.yml
Comment (cevich) [inline .cirrus.yml:744]: No need to define an ARM-specific anchor for this (since the actual content is identical).   Unless, were you intending to include "(ARM64)" in the string?

File: .cirrus.yml
Comment (cevich) [inline .cirrus.yml:759]: Likely we don't need an arm-specific anchor for this section.  Should be good enough to `always: *logs_artifacts`

File: .cirrus.yml
Comment (cevich) [inline .cirrus.yml:835]: Arguable this one could probably be skipped for arm.

**Comment by cevich** [comment]
@lsm5 you're probably going to need to rebase-onto and/or wait for https://github.com/containers/podman/pull/14972 as well.  Alternatively, you could **_temporarily_** disable (`skip: $CI == $CI`) all/most non-arm tasks so you can test the new tasks until that PR merges.

File: .cirrus.yml
Comment (lsm5) [reply .cirrus.yml:759]: ack

File: .cirrus.yml
Comment (lsm5) [reply .cirrus.yml:255]: removing VM_IMAGE_NAME..

File: .cirrus.yml
Comment (lsm5) [reply .cirrus.yml:835]: skipping rootless system tests

**Comment by lsm5** [comment]
`bud podman fedora-36 root host` looks like a redhat registry problem?

**Comment by lsm5** [comment]
@cevich @edsantiago can I get some eyes on `sys podman fedora-36 {aarch64} root host` ? Some command not found, don't know which one. And earlier Ed said `run -127` as indicated in the log suggestion is wrong.

**Comment by cevich** [comment]
@lsm5 I'm seeing a TON of failures due to the localhost registry being inaccessible.  I'm guessing this is due to the ARM64-image being inaccessible?  

Also: 
`# Error: error creating cgroup path /libpod_parent/7c44fa42363ca7d85fc935daa45b5c409ccd125ffe0d0eb56fe8d79116416502: write /sys/fs/cgroup/libpod_parent/cgroup.subtree_control: no such file or directory`

That path seems very wrong.  I haven't looked at the test-code, but that strikes me as something is very wrong with the environment (in general).

And:

`podman pull quay.io/libpod/fedora:31`

Best check if that is a manifest-list containing ARM64 image or just a regular x86_64 image.

And:

```
[+1446s] # system_u:system_r:cloud_init_t:s0^@
[+1446s] # #/vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv
[+1446s] # #|     FAIL: SELinux type
[+1446s] # #| expected: 'spc_t'
[+1446s] # #|   actual: 'cloud_init_t'
[+1446s] # #\^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

That's VERY broken.  What the heck is cloud-init doing touching any of this stuff?  cloud-init should run once at VM startup, then get out of the way.  No configuration I'm aware of should cause it to touch anything container-related like this.  Maybe SELinux policy is broken in some way for ARM?

**Comment by rhatdan** [comment]
I would guess you have a labeling issue.  Perhaps you need to relabel files if you are moving them around.

**Comment by cevich** [comment]
> I would guess you have a labeling issue. Perhaps you need to relabel files if you are moving them around.

Superficially sure, but if the test is passing w/o this error on `x86_64`, it suggests a problem specific to the `arm64` environment.  The images are built with substantially the same scripts, so that's why I wonder about a packaging problem.  Note: We **do not** touch `/.autorelabel` in the images, since that would worsen startup time.

**Comment by lsm5** [comment]
> > I would guess you have a labeling issue. Perhaps you need to relabel files if you are moving them around.
> 
> Superficially sure, but if the test is passing w/o this error on `x86_64`, it suggests a problem specific to the `arm64` environment. The images are built with substantially the same scripts, so that's why I wonder about a packaging problem. Note: We **do not** touch `/.autorelabel` in the images, since that would worsen startup time.

i rebased on main and i'm rerunning them now. I'll try getting terminal access on the next failure. 

@cevich RE: packaging problem, you mean, container-selinux / selinux-policy or something else?

**Comment by lsm5** [comment]
> @cevich @edsantiago can I get some eyes on `sys podman fedora-36 {aarch64} root host` ? Some command not found, don't know which one. And earlier Ed said `run -127` as indicated in the log suggestion is wrong.

so this failure is back and consistent on all three tests: 1 for x86_64 and 2 for aarch64.

@cevich @edsantiago PTAL

**Comment by edsantiago** [comment]
The [normal-arch fail](https://api.cirrus-ci.com/v1/artifact/task/6329943377313792/html/sys-podman-fedora-36-rootless-host.log.html) is almost certainly flake #10825, a pretty nasty one that's over a year old by now.

The new-arch ones are expected. It will take time to solve or skip those. I'll look into them one by one. If you need to merge this urgently, I suggest filing issues for each one, then adding `skip "FIXME: #12345"` to each one so the 12345s can be tracked.

**Comment by edsantiago** [comment]
Correction: `skip_if_aarch64`, which needs to be a new function (easy to write).

Why are the test name (and log files) including curly braces? Is that a deliberate design decision for these tests? It looks odd and seems to break some of my greasemonkey code. (Not a big deal, I can probably fix it, just curious)

**Comment by lsm5** [comment]
> Correction: `skip_if_aarch64`, which needs to be a new function (easy to write).

ack, will check and add
 
> Why are the test name (and log files) including curly braces? Is that a deliberate design decision for these tests? It looks odd and seems to break some of my greasemonkey code. (Not a big deal, I can probably fix it, just curious)

@cevich remind me please why you preferred {aarch64} in curly braces :)

File: .cirrus.yml
Comment (edsantiago) [inline .cirrus.yml:31]: Request: `-` (dash) instead of space? I suppose I can live with the curly braces, but I can't live with whitespace. Thanks.

**Comment by edsantiago** [comment]
`login` test failures seem to be because the registry image isn't multiarch -- but the 2.8 one is:
```console
$ buildah manifest inspect quay.io/libpod/registry:2.8 | jq -r '.manifests[].platform' 
... amd64, arm, arm64, etc
```
Quick suggestion (which I have no way to test):
```diff
diff --git a/test/system/150-login.bats b/test/system/150-login.bats
index dc902d5fe..b57bb44ab 100644
--- a/test/system/150-login.bats
+++ b/test/system/150-login.bats
@@ -52,7 +52,7 @@ function setup() {                                                                                                                           
     mkdir -p $AUTHDIR
 
     # Registry image; copy of docker.io, but on our own registry
-    local REGISTRY_IMAGE="$PODMAN_TEST_IMAGE_REGISTRY/$PODMAN_TEST_IMAGE_USER/registry:2.7"
+    local REGISTRY_IMAGE="$PODMAN_TEST_IMAGE_REGISTRY/$PODMAN_TEST_IMAGE_USER/registry:2.8"
 
     # Pull registry image, but into a separate container storage
     mkdir -p ${PODMAN_LOGIN_WORKDIR}/root
```

File: .cirrus.yml
Comment (lsm5) [reply .cirrus.yml:31]: i'm cool with it but I'll defer to @cevich.

**Comment by edsantiago** [comment]
[this one](https://api.cirrus-ci.com/v1/artifact/task/5555887191359488/html/sys-podman-fedora-36%20%7Baarch64%7D-root-host.log.html#t--00264) is in code that @cdoern just merged last week:
```
# # podman --cgroup-manager=cgroupfs pod create --name=resources-cgroupfs --cpus=5 --memory=5m --memory-swap=1g --cpu-shares=1000 --cpuset-cpus=0 --cpuset-mems=0 --device-read-bps=/dev/loop0:1mb --device-write-bps=/dev/loop0:1mb --blkio-weight-device=/dev/loop0:123 --blkio-weight=50
# Error: error creating cgroup path /libpod_parent/e0024c8b8ccc24c247b62a422433c0b69d7c3f930bad3863563fcec0d0db43f1: write /sys/fs/cgroup/libpod_parent/cgroup.subtree_control: no such file or directory
# [ rc=125 (** EXPECTED 0 **) ]
```

**Comment by edsantiago** [comment]
[this one](https://api.cirrus-ci.com/v1/artifact/task/5555887191359488/html/sys-podman-fedora-36%20%7Baarch64%7D-root-host.log.html#t--00291) is systemd, so @vrothberg might be the best person to look at it, but it also could be crun, so, ping, @giuseppe also:
```
# # podman run -d --sdnotify=container quay.io/libpod/fedora:31 sh -c printenv NOTIFY_SOCKET;echo READY;systemd-notify --ready;while ! test -f /stop;do sleep 0.1;done
# 2ff76f9670f13c479196440ac93babe9fc4afa8cbb0e0b6799b73a3b59969292
# # podman logs 2ff76f9670f13c479196440ac93babe9fc4afa8cbb0e0b6799b73a3b59969292
# /run/notify/notify.sock
# READY
# [0;1;31mFailed to notify init system: Permission denied[0m
```
All the permission and SELinux errors make me strongly suspect that SELinux is broken on these systems. It might be that the only way to debug is to ssh into one of them.

**Comment by cdoern** [comment]
@edsantiago what changed in this PR that could trigger a test to fail that passed 3 days ago? these things keep sneaking by CI mysteriously...

@giuseppe WDYT on the subtre_control failure? I have only seen that in rootless since we do not create libpod_parent there.

**Comment by edsantiago** [comment]
@cdoern this is a new arch.

File: .cirrus.yml
Comment (cevich) [reply .cirrus.yml:31]: Yes please, what Ed said, not having whitespace is probably important.

File: .cirrus.yml
Comment (cevich) [inline .cirrus.yml:37]: On your next update: please fix comment with ambiguous image IDs.  Also (ideally) all these images should come from the same build (same ID).  That lowers the amount of brain-power required when reviewing test results.  i.e. the `crun` version is the same in everything called Fedora.

Comment (lsm5) [reply .cirrus.yml:37]: one of the commits from unmerged PRs had the `matches: foo` thing, while sometime back you mentioned adding the cirrus task ID next to the ami. Let me know which one you prefer.

RE: fedora, is the current one showing different crun versions?

**Comment by cevich** [comment]
> trigger a test to fail that passed 3 days ago?
> @cdoern this is a new arch.

Actually maybe also: The images (AMI's) here (I think) were built *before* the latest `c6013173500215296` ones from last week.  I can't (easily) tell from the ambiguous comment.

Ugh...these unreadable/disconnected AMI ID's are going to drive us crazy.  I'm updating the tags on the AMI to point at the image ID, but there's no visibility of that in Cirrus or at runtime :confounded: We're literally relying on a comment to indicate which build they came from :angry:   I really don't know a good way to make this more foolproof :disappointed:

File: .cirrus.yml
Comment (lsm5) [reply .cirrus.yml:37]: and ya, I haven't updated the build ids today

Comment (cevich) [reply .cirrus.yml:37]: Definitely reference either the `c/automation_images` build ID or the `$IMAGE_SUFFIX` (basically the same thing, but with a `c` in front).  I opened a Cirrus-CI RFE to help make this easier for humans, but until then we'll need to rely on comments :disappointed:

Comment (cevich) [reply .cirrus.yml:37]: In case it makes your life easier, [these ***just*** built, feel free to use them:](https://github.com/containers/automation_images/pull/157#issuecomment-1194567684)

|Name|ID|
|---|---|
|**build-push**|`build-push-c5495735033528320`|
|**fedora**|`fedora-c5495735033528320`|
|**fedora-aws**|`ami-0df5df528071f1052`|
|**fedora-netavark**|`fedora-netavark-c5495735033528320`|
|**fedora-netavark-aws-arm64**|`ami-00c26c58f6eaba850`|
|**fedora-podman-aws-arm64**|`ami-02ee8b3a782a78791`|
|**fedora-podman-py**|`fedora-podman-py-c5495735033528320`|
|**prior-fedora**|`prior-fedora-c5495735033528320`|
|**ubuntu**|`ubuntu-c5495735033528320`|

File: test/system/helpers.bash
Comment (edsantiago) [inline test/system/helpers.bash:383]: In shell languages `-eq` is for numbers and `=` for strings. (I know. Please don't blame me.) Suggestin:
```bash
    [[ "$(uname -m)" == "aarch64" ]]
```
And, you actually have to use the new `skip_if_aarch64` code somewhere :)

File: test/system/200-pod.bats
Comment (edsantiago) [inline test/system/200-pod.bats:481]: Dup, please remove

File: test/system/helpers.bash
Comment (lsm5) [reply test/system/helpers.bash:383]: lol yes, i realized after pushing the image ID change that I had this diff too. I'll check out the failures and then add this.

File: .cirrus.yml
Comment (lsm5) [reply .cirrus.yml]: @cevich thanks, as part of the RFE, if possible, would be nice to have this table constantly updated on some fixed bookmarkable link so we don't have to look at PRs or logs.

**Comment by lsm5** [comment]
@edsantiago aarch64 no longer in curly braces. Added skip_if_aarch64 to a couple of tests that failed in the last run. Also, created issue #15074 for that.

File: test/system/075-exec.bats
Comment (edsantiago) [inline test/system/075-exec.bats:9]: Reliably??? Like, it fails consistently? Is it the `137` bug? If so, this could be an awesome opportunity to identify and fix #10825.

File: test/system/030-run.bats
Comment (edsantiago) [inline test/system/030-run.bats:6]: I don't remember this one failing. What was the symptom? (sigh, logs are lost)

**Comment by edsantiago** [comment]
I notice you didn't skip any of the SELinux tests. Did that get resolved with the new images?

File: test/system/075-exec.bats
Comment (lsm5) [reply test/system/075-exec.bats:9]: The system tests were failing similarly but I didn't keep track of which exact ones. Will do that going fwd. I can remove this change in that case and repush. Also, identifying and fixing are probably best left for later or for someone else cause my main focus is to get machine tests enabled in https://github.com/containers/podman/pull/15031 which depends on this one.

**Comment by lsm5** [comment]
> I notice you didn't skip any of the SELinux tests. Did that get resolved with the new images?

the last failure only showed the system test failures and nothing else (i guess others were cancelled). I'll see how this run goes.

**Comment by edsantiago** [comment]
Next time you rebase, please remove the 761e9f256 commit; and I think you can remove 8d9b7cac2 and 6b5c46735 as well

**Comment by lsm5** [comment]
> Next time you rebase, please remove the [761e9f2](https://github.com/containers/podman/commit/761e9f256b7461a8b43b4214f8fa3c66ad22b370) commit; and I think you can remove [8d9b7ca](https://github.com/containers/podman/commit/8d9b7cac29ad735b1d65ea57e0164036104c2bc0) and [6b5c467](https://github.com/containers/podman/commit/6b5c4673524f565467eb7f00dd9969fd5d2e88d6) as well

@edsantiago do we not need them anymore or do you need the removal to investigate something else?

**Comment by edsantiago** [comment]
All those have merged, although in slightly different form than when you rebased on them. The code that merged is Truth, and should not be clobbered by older versions.

**Comment by lsm5** [comment]
repushed and skip_if_aarch64 commented out to check for failure consistency.

**Comment by lsm5** [comment]
> > I notice you didn't skip any of the SELinux tests. Did that get resolved with the new images?
> 
> the last failure only showed the system test failures and nothing else (i guess others were cancelled). I'll see how this run goes.

the system test failures look consistent. It's the same 2 tests failing for remote and non-remote.

File: .cirrus.yml
Comment (cevich) [inline .cirrus.yml:742]: Minor nit: The last part of this comment is wrong, you're not adding the full compliment of system-test flavors (which is fine).  No need to re-run all the tests just for this change, but if you do have to re-push, I'd suggest maybe just nuke all/most of this comment block.

File: .cirrus.yml
Comment (cevich) [reply .cirrus.yml:744]: Minor nit: The `std_name_fmt_aarch64` anchor isn't referenced anywhere, you can remove it.  Also, I think since you're defining `DISTRO_NV` to include "aarch64", I think you can just use the `*std_name_fmt` alias here, no need to redefine.

File: .cirrus.yml
Comment (lsm5) [reply .cirrus.yml]: done.

File: .cirrus.yml
Comment (cevich) [inline .cirrus.yml:1120]: I'd prefer you just leave out the aarch64 artifacts task.  Let's wait until there's an actual need.  At that time, it should be included into the single common artifacts task (i.e. not a separate one) - to make downstream integration easier.

File: .cirrus.yml
Comment (cevich) [inline .cirrus.yml:1192]: This task isn't strictly needed.  Really it's just checking `version.go` details, which is not architecture or platform dependent.  I'd recommend just leaving this whole section out.

File: contrib/cirrus/cirrus_yaml_test.py
Comment (cevich) [inline contrib/cirrus/cirrus_yaml_test.py:32]: This isn't right.  I think the `test_image_build_aarch64` maybe shouldn't be here.  i.e. that task should be required for success.

The other additions are reasonable - unless/until you remove `artifacts_aarch64` and `release_test_aarch64` from `.cirrus.yml` as I commented.

File: contrib/cirrus/runner.sh
Comment (cevich) [inline contrib/cirrus/runner.sh:41]: This is odd, is there a difference here in our image build-scripts for aarch64?

**Review by cevich — APPROVED** [review]
Overall this is a fantastic effort that should be commended.  I made a few comments that should be addressed, but overall they're all pretty minor.  LGTM with some minor fixups and assuming tests pass.

File: contrib/cirrus/runner.sh
Comment (lsm5) [reply contrib/cirrus/runner.sh:41]: might have been a case of me earlier using the wrong image, let me get rid of that in the next push. I need to skip some other aarch64 tests I think

File: contrib/cirrus/cirrus_yaml_test.py
Comment (lsm5) [reply contrib/cirrus/cirrus_yaml_test.py:32]: so, test_image_build isn't required for success, but the aarch64 one is?

File: .cirrus.yml
Comment (lsm5) [reply .cirrus.yml:1120]: removed artifacts_aarch64.

File: contrib/cirrus/cirrus_yaml_test.py
Comment (lsm5) [reply contrib/cirrus/cirrus_yaml_test.py:32]: removed release_test_aarch64 and artifacts_aarch64 from .cirrus.yml

File: .cirrus.yml
Comment (lsm5) [reply .cirrus.yml:1192]: removed.

File: .cirrus.yml
Comment (lsm5) [reply .cirrus.yml]: done.

**Review by cevich — APPROVED** [review]
/lgtm
/hold

**Comment by edsantiago** [comment]
```
        _ruamel_yaml.c:6:10: fatal error: Python.h: No such file or directory
            6 | #include "Python.h"
              |          ^~~~~~~~~~
```

**Comment by cevich** [comment]
Oof, WTF.  Doing some digging...

**Comment by edsantiago** [comment]
nonono, that's just because @lsm5 tried removing something, but I guess it can't be removed. Should be easy to just add it back.

File: .cirrus.yml
Comment (cevich) [inline .cirrus.yml:398]: get rid of this line (sorry I missed it).  You're using an `ec2_instance` for this task, not the container image.  But I don't think that explains fully the missing `Python.h`.  Still poking around.

**Comment by cevich** [comment]
> nonono, that's just because @lsm5 tried removing something, but I guess it can't be removed. Should be easy to just add it back.

Our VM images should be uniform across architectures.  Since we're not seeing this error on the x86 tasks, something is clearly not uniform.

Looking in c/automation_images, `cache_images/fedora_packaging.sh` we don't mention `python-devel` anywhere.  I'm taking a quick poke around to see if I can find where it gets brought.

**Comment by cevich** [comment]
Okay, it looks like the `python-devel` is missing from the regular images too.  If you want to add back in the `dnf install` for those two tasks (Validate and consistency?) I think there's no need to check the architecture, just always install it.  Maybe drop a `TODO` comment suggesting we bundle it into the images for stability.  But I don't see that work as critical ATM.

**Comment by cevich** [comment]
Opened https://github.com/containers/automation_images/issues/159

**Comment by lsm5** [comment]
@edsantiago @cevich removed that smallcontainer line and added python3-devel back. PTAL once again please, sorry about the PITA :(

**Review by cevich — APPROVED** [review]
/lgtm
/hold

**Comment by cevich** [comment]
> sorry about the PITA :(

It's no problem bud.

**Comment by edsantiago** [comment]
[remote is a flake](https://api.cirrus-ci.com/v1/artifact/task/4683338345611264/html/sys-remote-fedora-36-root-host.log.html#t--00207) which should go away as soon as #15090 merges.

**Comment by edsantiago** [comment]
LGTM but that `resource` flake is going to cause a lot of pain. I foresee an immediate followup PR to `skip` it.

Thanks for your perseverance @lsm5.

**Review by vrothberg — COMMENTED** [review]
/lgtm

Thanks everybody for working on this. I am very excited!

**Pull request merged**
